git-remote-ipfs/README.rst

```diff
@@ -0,0 +1,153 @@
+communicated out-of-band. The output of `git remote get-url` can be used to `git
+clone`.
+
+An alternative is to use IPNS_, which provides a stable name the remote helper
+can update whenever the remote refs are updated:
+
+.. _IPNS: https://docs.ipfs.io/guides/concepts/ipns/
+
+.. code:: shell
+
+    $ repoid=$(ipfs key gen --type=ed25519 myrepo)
+    $ # Publish a pointer to the empty directory initially
+    $ ipfs name publish --key $repoid QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn
+    $ git remote add ipns ipfs://ipns/$repoid
+
+Note that resolving and updating an IPNS name is rather slow on the main IPFS
+network. Also note that only the owner of the IPNS name (that is, the keypair)
+can update it.
+
+Configuration
+--------------------------------------------------------------------------------
+
+A number of options regarding IPFS can be configured via git-config_:
+
+* ``ipfs.apiurl``
+
+  The URL of the IPFS daemon, default: ``http://localhost:5001``
+
+* ``ipfs.maxconnections``
+
+  The max number of connections to open to the IPFS daemon, default: ``30``
+
+* ``ipfs.maxblocksize``
+
+  The maximum block size (in bytes) supported by bitswap. Most users will
+  **not** want to change this. Default: ``2048000``
+
+The API URL can be overriden per-remote using the key
+``remote.<remote name>.ipfsapiurl``, e.g.:
+
+.. code:: shell
+
+    $ git config remote.origin.ipfsapiurl http://127.0.0.2:5001
+
+Additionally, if the environment variable ``IPFS_API_URL`` is set, it will be
+used instead of any git-config_ settings.
+
+.. _git-config: https://git-scm.com/docs/git-config
+
+How it works
+================================================================================
+
+IPFS blocks can be created with the ``git-raw`` CID format, which allows IPFS to
+interpret the data as loose git objects. When created with ``sha1`` as the
+(multihash) hash function, the block's CID corresponds to the SHA1 hash of the
+git object, i.e. one can be recovered from the other. Crucially, this allows the
+SHA1 references embedded in a loose git object (eg. parents and tree of a
+commit) to be traversed given a head reference.
+
+In order to obtain the head reference, IPLD links are created corresponding to
+the ``refs/heads`` directory hierarchy. Note that adding links to an IPFS object
+changes its hash - this means each push results in a new object (CID),
+which must be retained in order to clone or pull.
+
+Which git objects need to be pushed or fetched is determined via the `git remote
+helper`_ protocol, respectively by inspecting the local git repo and remote
+refs.
+
+Limitations
+--------------------------------------------------------------------------------
+
+* It is currently unclear how to keep track of the latest "anchor" object (the
+  one linking to the most recent heads). The obvious solution is to to use IPFS'
+  native name resolution mechanism (IPNS), yet IPNS names have a very limited
+  lifetime on the main IPFS network.
```

Comment:
Maybe we could add a link to this discussion: https://discuss.ipfs.io/t/ipns-max-lifetime/2130